File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_4_isp/after/Goalie.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_4_isp.after;

/**
 * Interface Composition.
 * * ARCHITECTURE NOTE: The beauty of Interface Segregation is that classes 
 * can just "opt-in" to the behaviors they actually need. Because the Goalie 
 * needs to practice everything, it simply implements BOTH interfaces. 
 * * We accommodated the complex requirements of the Goalie without polluting 
 * the Midfielder's codebase!
 */
public class Goalie implements FieldPlayerTraining, GoalieTraining {
    
    public void practiceShooting() {
        System.out.println("  [Goalie] Practicing goal kicks and long shots.");
    }

    public void practiceTackling() {
        System.out.println("  [Goalie] Practicing 1-on-1 box tackles.");
    }

    public void practiceDivingSaves() {
        System.out.println("  [Goalie] Practicing top-corner diving saves.");
    }

    public void practiceHandDistribution() {
        System.out.println("  [Goalie] Practicing fast break throws.");
    }
}

File: Chapter 03/Python/section_3_3_4_isp/after/goalie.py
from .interfaces import FieldPlayerTraining, GoalieTraining

"""
INTERFACE COMPOSITION.
* ARCHITECTURE NOTE: Python handles ISP beautifully through multiple 
inheritance. Because a Goalie needs both general field skills and 
specialized net skills, we simply compose the Goalie out of both 
interfaces. 
* This provides total flexibility without polluting the base definitions 
for other players.
"""
class Goalie(FieldPlayerTraining, GoalieTraining):
    def practice_shooting(self):
        print("  [Goalie] Practicing goal kicks and long shots.")

    def practice_tackling(self):
        print("  [Goalie] Practicing 1-on-1 box tackles.")

    def practice_diving_saves(self):
        print("  [Goalie] Practicing top-corner diving saves.")

    def practice_hand_distribution(self):
        print("  [Goalie] Practicing fast break throws.")